controllers/network/openstackipset_controller.go

```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers. Return and don't requeue.
+			return result, nil
+		}
+		return result, err
+	}
+
+	helper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		r.Log,
+	)
+
+	if err != nil {
+		// helper might be nil, so can't use util.LogErrorForObject since it requires helper as first arg
+		r.Log.Error(err, fmt.Sprintf("unable to acquire helper for OpenStackIPSet %s", instance.Name))
+		return result, err
+	}
+
+	defer func() {
+		if err != nil {
+			instance.Status.Conditions.MarkTrue(networkv1beta1.OpenStackIPSetReadyCondition, networkv1beta1.OpenStackIPSetReadyMessage)
+		}
+		// update the overall status condition if service is ready
+		err := helper.PatchInstance(ctx, instance)
+		if err != nil {
+			r.Log.Error(err, "Can't patch Instance")
+			_err = err
+			return
+		}
+	}()
+
+	if instance.DeletionTimestamp.IsZero() && controllerutil.AddFinalizer(instance, helper.GetFinalizer()) {
+		return ctrl.Result{}, err
+	}
+
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.ReconcileDelete(ctx, instance, helper, req)
+	}
+
+	instance.InitCondition()
+
+	// Build AssignIPDetails
+	networks := instance.Spec.Networks
+
+	if instance.Status.IPAddresses == nil {
+		instance.Status.IPAddresses = make(map[string]networkv1beta1.IPReservation)
+	}
+
+	//find the the openstacknet objects
+	for _, network := range networks {
+		if _, ok := instance.Status.IPAddresses[network.Name]; ok {
+			continue
+		}
+		osnet := &networkv1beta1.OpenStackNet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      network.Name,
+				Namespace: instance.Namespace,
+			},
+		}
+		err = r.Get(ctx, types.NamespacedName{Name: network.Name, Namespace: instance.Namespace}, osnet)
+		if err != nil {
+			r.Log.Error(err, "Network Not Found")
+			return ctrl.Result{}, err
+		}
+		aid := ipam.AssignIPDetails{}
+		for i, subnet := range osnet.Status.Subnets {
+			if subnet.Name == network.SubnetName {
+				_, cidr, err := net.ParseCIDR(subnet.Cidr)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+				aid.IPNet = *cidr
+
+				aid.Reservelist = subnet.Reservations
+				aid.RangeStart = net.ParseIP(subnet.AllocationRange.AllocationStart)
+				aid.RangeEnd = net.ParseIP(subnet.AllocationRange.AllocationEnd)
+				ipReservation, updatedReservations, err := ipam.AssignIP(aid, net.ParseIP(network.FixedIP))
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+				ipReservation.Gateway = subnet.Gateway
+				ipReservation.Routes = subnet.Routes
+
+				aid.Reservelist = updatedReservations
+				osnet.Spec.Subnets[i].Reservations = append(osnet.Spec.Subnets[i].Reservations, ipReservation)
+
+				err = r.Update(ctx, osnet)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
```

Comment:
I think we want to use Patch() instead of update that on reconciliation the osnet only gets updated if this controller did a change to the current object version.

Reply:
I'll check if patch works and update accordingly.

Reply:
if it does not work, we should use deepequal to check if there was a change and only call the update if there is a change

Reply:
I changed to use patch, but did not notice any benefits and we've already got checks to not update when already reserved.